### PR TITLE
Implement Sysvar Trait for Instruction introspection (Incomplete attempt)

### PR DESCRIPTION
**Pull request by k--chow**
#### Problem
There are multiple issues opened about implementing the Sysvar trait for Instructions. 
- https://github.com/solana-labs/solana/issues/22911
- https://github.com/solana-labs/solana/issues/17017
- https://github.com/solana-labs/solana/issues/23617

Presently, this account must be passed in and was involved in the wormhole hack (was not checked properly), and is needed to check signatures with the Secp256k1 program (one reason why the Secp256k1 is hard to use).

Anchor also must use `UncheckedAccount` forcing an address=id() right now.

#### Stuck
From https://docs.solana.com/implemented-proposals/instruction_introspection:
"The runtime will recognize this special instruction, serialize the Message instruction data for it and also write the current instruction index and then the bpf program can extract the necessary information from there."

I've tried to follow along with replicating from `Clock::get()`, but I'm missing some key piece to do this. 

## Update
The runtime checks the passed in accounts, and builds the `instruction sysvar` if it is passed in. However, it is not in anywhere global and is only stored in the accounts of `LoadedTransaction`.
Source: https://github.com/solana-labs/solana/blob/c08cfafd6c599abbf440fbac1974ac8d754e265f/runtime/src/accounts.rs#L265-L270

# Proposal
1. Create an Instruction Introspection struct inside `sdk/program/src` (where Clock and SlotHistory are)
2. when `load_accounts` is called in `bank.rs`, update Instruction Introspection (like how Clock or SlotHistory is updated) to always hold the transactions currently being processed. Help needed on whether this is exorbitant overhead (since this computation normally only happens if instruction sysvar is passed in)

Code for how the SlotHistory sysvar is updated
https://github.com/solana-labs/solana/blob/c08cfafd6c599abbf440fbac1974ac8d754e265f/runtime/src/bank.rs#L2303-L2315
4. Now implement the Sysvar Trait. Update the methods like `load_instruction_at_checked` to be methods of Instruction Introspection and not the sysvar as it is now. The struct now acts similar to Clock and can be "get()"d.

## Pros
- can be more organized and can clean up the hacky way Sysvar implements and deserializes the data
## Cons/Questions
- more overhead computation since called in every transaction even if not needed (unless there's another way to indicate the instruction needs it?). Maybe the runtime can detect if the instruction `gets()` on instruction sysvar?
- `load_accounts` in `accounts.rs` loads a batch of transactions, when is the right time to update the instruction sysvar to the current transaction?

Thoughts on this?

@armaniferrante @jarry-xiao

### CODE REVIEW

File: sdk/program/src/sysvar/instructions.rs [hunk 1]
```diff
@@ -119,7 +129,8 @@ pub fn load_current_index_checked(
     if !check_id(instruction_sysvar_account_info.key) {
         return Err(ProgramError::UnsupportedSysvar);
     }
-
+    let instruction_sysvar = solana_program::sysvar::instructions::try_borrow_data()?;
```
Comment (k--chow) [inline sdk/program/src/sysvar/instructions.rs:132]: This doesn't work

File: sdk/program/src/sysvar/instructions.rs [hunk 2]
```diff
@@ -119,7 +129,8 @@ pub fn load_current_index_checked(
...
+    let a = Instructions::from_account_info(instruction_sysvar_account_info).unwrap().try_borrow_data()?;
```
Comment (k--chow) [inline sdk/program/src/sysvar/instructions.rs:133]: Ignore this, playing around with it

File: sdk/program/src/sysvar/instructions.rs [hunk 3]
```diff
@@ -2,30 +2,40 @@
...
+
 #[cfg(not(target_arch = "bpf"))]
 use {
     crate::serialize_utils::{append_slice, append_u16, append_u8},
     bitflags::bitflags,
 };
 
 // Instructions Sysvar, dummy type, use the associated helpers instead of the Sysvar trait
+#[derive(Serialize, Clone, Deserialize, Debug, Default, PartialEq)]
 pub struct Instructions();
```
Comment (k--chow) [inline sdk/program/src/sysvar/instructions.rs:22]: I can't wrap my head around how instruction introspection is working (apparently it does something extra at runtime?). It seems to be more complicated than implementing a trait to the dummy struct.

**Comment by t-nelson** [comment]
I think we actually want to move away from the account-based sysvar access path, instead preferring the syscall based one.  So this may not be a desired contribution.

@jackcmay care to weigh in?

**Comment by jackcmay** [comment]
> I think we actually want to move away from the account-based sysvar access path, instead preferring the syscall based one. So this may not be a desired contribution.
> 
> @jackcmay care to weigh in?

Passing sysvars as accounts complicates composibility and we would like to move away from passing them as accounts and toward obtaining them via `.get()`.  Unfortunatly not all sysvars are conducive to this.  I'll take a look at adding support for `.get` to the instruction sysvar

**Comment by k--chow** [comment]
@t-nelson @jackcmay 
Thanks for the response. I believe that is what this PR is trying to accomplish here (implement sysvar trait so .get() is available through a system call) but I'm a bit stuck or it may not be conducive to do so as you say.

**Comment by jackcmay** [comment]
@k--chow Would `get_processed_sibling_instructions` work for the use cases you had in mind?
https://github.com/solana-labs/solana/blob/2a5764ef79cff391da080cc19617f171109c4158/sdk/program/src/instruction.rs#L684

We could implement `.get` to return the account data and then have the caller extract which instruction from there (aka, copy the account data into the program space) just like they are doing with the account data presently.  This does mean allocating space for the data from the heap.

**Comment by jackcmay** [comment]
The `Sysvar` trait doesn't really play nice because it assumes you are deserializing entirely to a specific type whereas the `Instructions` sysvar's "result" is an instruction at a particular index or the current index.

**Pull request closed without merging**